Rollup of all changes made for 1.6.1 stock

linuxgurugamer opened the pull request:
1.	Replaced all commented out 1.4 code with #if KSP1_4/#endif blocks
2.	Changed the /* 1.4 Addition   to   // 1.4 Addition
3.	Added coroutine:  HandleEditorButton_Coroutine() to reset the button 
	handlers every 1/2 second
4.	Added StartCoroutine to end of Start() in KerbalConstructionTime.cs

The following were done to optimize the code a bit, to remove unnecessary code from a FixedUpdate
(I saw a FIXME comment and decided to do so)

5.	Created new function 
		KerbalConstructionTime.UpdateTechlistIconColor() 
	which contains the code which was in the FixedUpdate()
6.	Commented out the code in FixedUpdate() which was replaced by new method
5.	Changed the TechList from a list to an IList, and in all methods which change the list, 
	added a call to 
		KerbalConstructionTime.instance.UpdateTechlistIconColor();
	to each method in the IList which changes the list.

More changes, related to release

6.  Renamed .version to KerbalConstructionTime.version
7.  Copied KerbalConstructionTime.version to root directory
8.  Renamed KerbalConstructionTime.csproj to KerbalConstructionTime-KSP1-4.csproj
9.  Put original KerbalConstructionTime.csproj back in place
10. Created stock copy of .version file
11. Copied the assemblyversion.tt to AssemblyVersion-Stock.tt
12. Modified AssemblyVersion-Stock.tt to use the Stock.version file

Additional:
Commented out the KCT_UpdateChecker, it was still looking ath the old magico13.net site

Some minor changes were made due to the Mono compiler on Linux being slightly older (?) than the VS compiler, changes were in KCT_Events.SubscribeToEvents and have been commented.  No functionality changed, just moved some inline code into seperate methods

MikeOnTea commented:
Hey LGG, thanks for the PR.
Looking briefly at it, i noticed two things. For one, you sometimes just comment out old code, sometimes adding a comment on why you did it. It's probably better to just remove old code to keep the code clean and readable. Why something was done can be said in the commit message that does the change, thus separating the code and its history.

Another minor thing, your TechItemIList should ideally be a generic ObservableList. An example of how that could look is here: https://answers.unity.com/questions/677956/getting-notified-when-modifying-a-collection-ala-c.html

siimav commented on Kerbal_Construction_Time/KCT_Utilities.cs:
```diff
@@ -1559,18 +1573,30 @@ public static bool RecoverActiveVesselToStorage(KCT_BuildListVessel.ListType lis
+
+                // Recovering the vessel in a coroutine was generating an exception insideKSP if a mod had added
+                // modules to the vessel or it's parts at runtime.
+#if false
```
What's going on with those if false preprocessor directives? As Mike suggested, old unused code shouldn't be just left laying around.

linuxgurugamer replied:
I didn't know if that was needed for the RO version.  If it isn't needed, then I'd be very happy to delete it

siimav commented:
I would also like to suggest a change to your script that generates the AssemblyInfo file. The problem is that if you change the AssemblyVersion attribute with each release, you're going to break compatibility with other mods that have a C# project reference to it. The CLR compiler will generate a static binding to the referenced assembly and loading it will then fail as soon as the AssemblyVersion of the actual DLL is different from the one used at compile-time. Therefore I would suggest keeping the AssemblyVersion attribute as a constant value and only incrementing it if you're making major changes that will anyway break compatibility with other mods. Instead of the AssemblyVersion you should increment the AssemblyFileVersion attribute with each release.
NK briefly mentioned that he added a new [KSPAssembly](https://github.com/KSP-RO/RP-0/blob/master/Source/Properties/AssemblyInfo.cs#L38) attribute that supposedly somehow solves this issue with AssemblyVersion changes. Unfortunately I haven't had the time to figured out how it works.

pap1723 commented:
I would second @siimav thoughts on the AssemblyVersion. I am constantly running into version issues and having to re-reference .DLL files, which then break others games if they are not using the exact same version. Often, these failures happen without any notice to the player and the mod just doesn't work. (There is reference in the log files, but that is more work for all the modders to look at and debug)

linuxgurugamer commented:
> Hey LGG, thanks for the PR.
> Looking briefly at it, i noticed two things. For one, you sometimes just comment out old code, sometimes adding a comment on why you did it. It's probably better to just remove old code to keep the code clean and readable. Why something was done can be said in the commit message that does the change, thus separating the code and its history.
> 
> Another minor thing, your TechItemIList should ideally be a generic ObservableList. An example of how that could look is here: https://answers.unity.com/questions/677956/getting-notified-when-modifying-a-collection-ala-c.html

All the work I do with this mod is being done with the idea that the changes will be pushed back into RO.  Because of that, I don't want to delete code that might be needed in the RO version.  If not needed, then it can obviously be removed.

Also, regarding old code, I've seen that many non-professional modders don't use Git to it's fullest advantage.  I comment things the way I do to make it more understandable for them.

linuxgurugamer commented:
In the ToolbarController , in the AssemblyInfo, it has the following

    [assembly: AssemblyFileVersion("1.0.0.0")]
    [assembly: KSPAssembly("ToolbarController", 1, 0)]

and in another file, generated from the .version file and created by my TT script, it has:

    [assembly: AssemblyVersion("0.1.6.20")]

I've changed version numbers many times, and never had any incompatibilities.  

The KSPAssembly is what I use to increment when the API changes.  The only time I've ever had to do that was with the ClickThroughBlocker, when I had to change that to support another mod.  In the ClickThroughBlocker, it has:

    [assembly: KSPAssembly("ClickThroughBlocker",1, 7, 0)]

and one mod (the old Blizzy toolbar) which needed the change has:

    [assembly: KSPAssemblyDependency("ClickThroughBlocker", 1, 7, 0)]


Other mods have this in them:

    [assembly: KSPAssemblyDependency("ToolbarController", 1, 0)]

which forces the dependency checking against the KSPAssembly.  This makes sure that the right version of ToolbarController is loaded, but the KSPAssembly never changes, so other mods don't have any issues with it.

linuxgurugamer commented:
BTW, the update I did this morning was to fix a nasty bug which caused problems when other mods added partmodules at runtime.  THAT update is not yet complete, I have a beta version and am waiting for feedback.  Once I get confirmation that it works on other computer, there are some commented code which I WILL be removing

linuxgurugamer commented:
Just an FYI, the TT and the AssemblyFile stuff is what I use to support all my mods.  While I obviously think it is useful, there is no requirement that it be used in the RO version.  If you decide you don't want it, I'll set it up so that the AssemblyInfo.cs will be as it was for RO, and for my version I'll have it, and I'll deal with any issues that arise.

linuxgurugamer commented:
> Another minor thing, your TechItemIList should ideally be a generic ObservableList. An example of how that could look is here: https://answers.unity.com/questions/677956/getting-notified-when-modifying-a-collection-ala-c.html

It took me a bit, but now I see the difference.  I'll update to use an ObservedCollection in a day or so

MikeOnTea commented:
The difference is that the TechItemList class is tied to KCT code because it contains these "KerbalConstructionTime.instance.UpdateTechlistIconColor();" calls in many places. A generic list, like that ObservedCollection, does not contain any KCT related code, can be put in its own file and potentially be reused in different places/projects. You'd just tell that list instance what to do on updates/changes by setting the action delegates, and you'd only have to put the UpdateTechlistIconColor() call in there once.

linuxgurugamer commented:
> The difference is that the TechItemList class is tied to KCT code because it contains these "KerbalConstructionTime.instance.UpdateTechlistIconColor();" calls in many places. A generic list, like that ObservedCollection, does not contain any KCT related code, can be put in its own file and potentially be reused in different places/projects. You'd just tell that list instance what to do on updates/changes by setting the action delegates, and you'd only have to put the UpdateTechlistIconColor() call in there once.

I've done that, it's in the current release.  And thanks for the explanation

siimav commented:
I have now managed to playtest your changes and verify that eveything is working correctly. So I'm almost ready to to merge this PR to the master.
After that gets done, I think I'll need to re-add the compiled dll to the repo. I don't really like these laying around either but people are used to downloading the 'master' branch to get the latest developmental version. We don't have any build servers so I can't currently think of a better way to solve this.
Also I would suggest a change to how assembly versioning is handled in this project. What I would like to do is to keep AssemblyVersion as a fixed value and only increment AssemblyFileVersion with each release. This would also require a tiny change to your t4 template. 

If you're interested, below is what I posted on Discord about how the assembly versioning works in KSP:
> Did some digging into KSP's AssemblyLoader code. It appears that it uses dynamic assembly loading (https://docs.microsoft.com/en-us/dotnet/framework/app-domains/resolve-assembly-loads) to get around the issue of not having the same assembly version that was used at compile-time. Basically that code just returns the assembly that was found from gamedata to Mono runtime and hopes that everything will work out. The caveat is that it will only do it when the version found from gamedata is larger than the one used at compile-time. So basically when I used the latest RF to compile RP0 dll but the user has some older version installed, bad things will start to happen.
NK also mentioned attributes called KSPAssembly and KSPAssemblyDependency but my understanding is that these are only used for dependency checking. Basically when RP0 sets version 1.1 of KCT as it's dependency but version 1.0 is found instead, then RP0 assembly will not get loaded.

linuxgurugamer commented:
That's very interesting.  I guess I've been saved by using AVC with everything.
The change to the tt is simple, it's just changing the attribute being written, to be AssemblyFileVersion instead of AssemblyVersion.
For reasons related to my process, I'll have to, for now, continue using AssemblyVersion in my mods, and, for the immediate future, my version of KCT.  But I'll make the changes so that the RO version will have AssemblyFileVersion updated.
I'll get this done tomorrow.

linuxgurugamer commented:
I just committed changes to address the issue of AssemblyVersion vs AssemblyFileVersion:
	Added AssemblyVersion-RO.tt for the RO build
	Modified AssemblyVersion-RO.tt to write AssemblyFileVersion instead of AssemblyVersion
	Commented out the AssemblyFileVersion in VersionInfo.cs
	Renamed KerbalConstructionTime.version to KerbalConstructionTime-RO.version
	Added Pre-build steps to run the TextTemplate
	Added a new directory called SpecialSurprise to hold code not being used for now

The only change I made to the RO .csproj was to add the prebuild steps.  The RO tt and RO .version files are specific to RO, and will write the AssemblyFileVersion as requested.  The Stock equivilents will be used my myself.  
The SpecialSurprise is something that was in KCT in 2015 for April's fools.  I'm not sure if I'm going to do anything with it, which is why I put it into a seperate directory.  No intention of having RO use it, currently it doesn't even compile, it's very old code.

siimav commented:
Huh, just noticed that the AssemblyVersion-Stock uses AssemblyFileVersion attribute but AssemblyVersion-RO has AssemblyVersion instead. I'm guessing it should be the other way around as you said that your build system currently need to increment AssemblyVersion.

linuxgurugamer commented:
Good catch.  Yes, they were, I just fixed it

linuxgurugamer commented:
Looks like I'll have to do another PR for this

linuxgurugamer commented:
Never mind, I see you did it already

linuxgurugamer commented:
> Hey LGG, thanks for the PR.
> Looking briefly at it, i noticed two things. For one, you sometimes just comment out old code, sometimes adding a comment on why you did it. It's probably better to just remove old code to keep the code clean and readable. Why something was done can be said in the commit message that does the change, thus separating the code and its history.
> 
> Another minor thing, your TechItemIList should ideally be a generic ObservableList. An example of how that could look is here: https://answers.unity.com/questions/677956/getting-notified-when-modifying-a-collection-ala-c.html

I'm working on my next release, this seems to be the only way for me to communicate with you guys.  The Issues functionality of Github is not enabled for this mod

As I'm working my way through this, I've found one or more large sections of "commented out code".  One which I'm looking at right now is in the KCT_BuildListStorage.cs, starting at line 64 and going through to line 108.  I've found various sections in various sizes throughout the code.

Is there a reason I'm not aware of that these are still in the code?

siimav commented:
That code you mentioned predates any our changes so I would assume that it's safe to delete those blocks.
I agree that the issues tab should be opened on this repo. Unfortunately I don't have the permissions to do that myself. @pap1723 ?
Also, if you want a more efficient way for discussing something with us, then maybe you should consider joining the [KSP-RO discord](https://discord.gg/3xMtyBg)?

linuxgurugamer commented:
I wasn't aware of the discord, thanks.  Will  join in a few minutes
